### DISCUSSION TITLE
Is there any function like AddScriptToExecuteOnDocumentCreated() in WebView2

### DISCUSSION THREAD

**Discussion by lk9015** [c-smile/sciter-js-sdk#100] [Q&A] [answered]
Is there any function like AddScriptToExecuteOnDocumentCreated() in WebView2 ?
Thanks in advance !

**Comment by c-smile** [comment, marked as answer]
There are multiple ways to achieve that. So context is required.

I am assuming that you will need that for some third party documents, if so then you should load such documents in &lt;frame&gt;s.

In Sciter you can define ```<frame content-style="content.css" src="...">```.  The frame is ordinary element similar to iframe in browsers - so it can appear in your root document (the one that is loaded into window directly).

That content.css may have this declaration:
```
html { 
   aspect: UsersContent url(.../users-content-handler.js);
} 
// some other CSS definitions for user's content like predefined components:
widget.file-selector {
   prototype: FileSelector url(.../users-content-handler.js);
   // other styles for it 
}
```

[Aspect in Sciter](https://sciter.com/sciter-declarative-behavior-assignment-by-css-prototype-and-aspect-properties/) is a plain JS function that will be called once when matching element appears in the DOM and so users-content-handler.js may look like as:
```
export function UsersContent() {
   // "this" here is the DOM element 
   ...
}
```
That users-content-handler.js will be loaded as module into any document loaded into the frame and you can do whatever you want in that script.

**Comment by lk9015** [comment]
Thanks for the response, 
AddScriptToExecuteOnDocumentCreated runs script before any other script, but the aspect doesn't.

```
<html>
<head>
    <style type="text/css">
    html { 
        aspect: initScript url("data:application/javascript,export function initScript(){ globalThis.xcall = Window.this.xcall}");
    } 
    </style>
    <script> 
    	console.log(xcall) // ReferenceError: 'xcall' is not defined
    </script>
</head>
```

**Reply by c-smile** [reply]
If you want to inject script _before_ any other scripts then this can be achieved (in script) only with frames:

Main document, loaded in window directly:
```html
<html>
    <head>
        <title>Test</title>
        <style></style>
        <script> 
          document.$("frame").on("parsed",function(evt) {
            let document = evt.target;
            document.globalThis.eval("xcall = Window.this.xcall");
          });
        </script>
    </head>
    <body>

<frame src="child.htm" />

    </body>
</html>
```
And child.htm
```html
<html>
    <head>
        <script>
         console.log(xcall);
        </script>
    </head>
    <body>
      Child
    </body>
</html>
```
In native code you can do that even for the root document by handling DOCUMENT_PARSED event.

**Reply by lk9015** [reply]
Thanks a lot, it works for me
